### PR TITLE
Refactor base classes to allow other implementations (JDBC)

### PR DESCRIPTION
**Pull request by ctron**

### CODE REVIEW

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonCredentials.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+package org.eclipse.hono.service.management.credentials;
+
+import java.util.Collection;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Helper methods for {@link CommonCredential}.
+ */
+public final class CommonCredentials {
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonCredentials.java:23]: how about simply adding these static methods to the existing *CommonCredential* class?

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/AutoProvisioning.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     */
+    public static Future<OperationResult<String>> provisionDevice(
+            final DeviceManagementService deviceManagementService,
+            final CredentialsManagementService credentialsManagementService,
+            final String tenantId,
+            final X509Certificate clientCertificate,
+            final Span span) {
+
+        Objects.requireNonNull(tenantId);
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/AutoProvisioning.java:65]: maybe we should add non-null checks for the services as well?

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractRegistrationService.java [hunk 3]
```diff
@@ -103,11 +104,39 @@ public final void setTenantInformationService(final TenantInformationService ten
...
      *            parent for any spans created in this method.
-     * @return A future indicating the outcome of the operation. A failed future with 
+     * @return A future indicating the outcome of the operation. A failed future with
      *           a {@link ServiceInvocationException} is returned, if there was an error resolving the groups.
      * @throws NullPointerException if any of the parameters is {@code null}.
      */
-    protected abstract Future<JsonArray> resolveGroupMembers(String tenantId, JsonArray viaGroups, Span span);
+    Future<JsonArray> resolveGroupMembers(final String tenantId, final JsonArray viaGroups, final Span span) {
+
+        final var viaGroupsAsString = viaGroups.stream()
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractRegistrationService.java:113]: how about adding null checks?

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonCredentials.java (hunk 1 above)
Comment (ctron) [reply services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonCredentials.java:23]: I was following the pattern of splitting actual objects and helper classes apart, like `java.util.Collections`, `java.lang.Objects`, `com.google.collect.Streams`, …

**Comment by ctron** [comment]
Added null checks.

**Review by kaniyan — CHANGES_REQUESTED** [review]

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java [hunk 4]
```diff
@@ -41,23 +43,34 @@
...
+    public AbstractCredentialsManagementService(
+            final Vertx vertx,
+            final HonoPasswordEncoder passwordEncoder,
+            final int maxBcryptIteration,
+            final Set<String> hashAlgorithmsWhitelist) {
+
         this.vertx = Objects.requireNonNull(vertx);
         this.passwordEncoder = Objects.requireNonNull(passwordEncoder);
+        this.maxBcryptIteration = maxBcryptIteration;
+        this.hashAlgorithmsWhitelist = hashAlgorithmsWhitelist;
```
Comment (kaniyan) [inline services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java:72]: Null check is missing for `hashAlgorithmsWhitelist`. `hashAlgorithmsWhitelist` is passed as an argumet to 
 `DeviceRegistryUtils.checkCredential(...)` and it does not allow _null_ value for this argument.

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsService.java [hunk 5]
```diff
@@ -32,8 +44,13 @@
  */
 public abstract class AbstractCredentialsService implements CredentialsService {
 
+    private static final Logger log = LoggerFactory.getLogger(AbstractCredentialsService.class);
```
Comment (kaniyan) [inline services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsService.java:47]: How about naming this static variable in upper case?

**Review by kaniyan — CHANGES_REQUESTED** [review]

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsService.java (hunk 5 above)
Comment (ctron) [reply services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsService.java:47]: First of all, it is not a constant, in the sense of a scalar value constant. Second, we use lowercase `log` in all other areas of Hono, so I would leave it as is.

**Comment by ctron** [comment]
Added javadoc about null and added "optional" handling for the allowed hash algorithm list

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java [hunk 6]
```diff
@@ -43,21 +46,31 @@
...
-    public AbstractCredentialsManagementService(final Vertx vertx, final HonoPasswordEncoder passwordEncoder) {
+    public AbstractCredentialsManagementService(
+            final Vertx vertx,
+            final HonoPasswordEncoder passwordEncoder,
+            final int maxBcryptIteration,
+            final Set<String> hashAlgorithmsWhitelist) {
+
         this.vertx = Objects.requireNonNull(vertx);
         this.passwordEncoder = Objects.requireNonNull(passwordEncoder);
+        this.maxBcryptIteration = maxBcryptIteration;
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java:71]: FMPOV we should also make sure that the iterations are within the supported value range, in our case [4, 31].
Having read throught the bcrypt documentation again, it seems like the parameter should better be called *maxBcryptLogRounds* as the number of actual iterations is computed as 2^logRounds ...

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonCredentials.java (hunk 1 above)
Comment (sophokles73) [reply services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonCredentials.java:23]: I see

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java (hunk 6 above)
Comment (ctron) [reply services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java:71]: I would say this is already covered by the configuration, which explicitly checks for that. Also would renaming mean, that we should do that consistently throughout the project. Not only here.

So I would say, we keep this PR focused on the current items. And create a new one later for renaming that parameter, which has an impact on documentation and configuration as well.

Comment (sophokles73) [reply services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java:71]: fine with me

**Review by sophokles73 — COMMENTED** [review]
this looks good to me. However, I do not think that we will be able to properly review the JDBC implementation in time for 1.4.0 so FMPOV there is no need to rush this into 1.4.0 as well. WDYT? Can this wait till 1.5.0?

**Comment by ctron** [comment]
> this looks good to me. However, I do not think that we will be able to properly review the JDBC implementation in time for 1.4.0 so FMPOV there is no need to rush this into 1.4.0 as well. WDYT? Can this wait till 1.5.0?

As long as we can agree on not merging anything which would break this. And that the release is still on for this week. Absolutely.

**Review by sophokles73 — APPROVED** [review]
LGTM

ready to be merged FMPOV

**Pull request merged**